File: brady/subtree_of_another_tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isSubtree(TreeNode root, TreeNode subRoot) {
        // find value at subRoot in root's tree using breadth-first-search
        Queue<TreeNode> bfsQueue = new LinkedList<>();
        bfsQueue.add(root);
        while (!bfsQueue.isEmpty()) {
            TreeNode node = bfsQueue.poll();
            if (node.val == subRoot.val && isSameTree(node, subRoot)) return true;

            // Adding nodes to the LinkedList is O(1) since Java LinkedLists are Deques
            if (node.left != null) bfsQueue.add(node.left);
            if (node.right != null) bfsQueue.add(node.right);
        }

        return false;
    }

    public boolean isSameTree(TreeNode p, TreeNode q) {
        if (p == null && q == null) return true;
        else if (p == null || q == null) return false;

        if (p.val != q.val) return false;

        return isSameTree(p.left, q.left) && isSameTree(p.right, q.right);
    }
}

File: gabe/subtree_of_another_tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

 #include <stack>

class Solution {
public:
    bool isSubtree(TreeNode* root, TreeNode* subRoot) {
        // find a possible root

        std::stack<TreeNode*> stck;
        TreeNode *current;

        stck.push(root);

        while(!stck.empty()) {
            current = stck.top();

            if (current->val == subRoot->val) {
                if (isSameTree(current, subRoot)) return true;
            }

            stck.pop();

            if (current->left != nullptr) stck.push(current->left);
            if (current->right != nullptr) stck.push(current->right);
        }

        return false;
    }



    bool isSameTree(TreeNode* p, TreeNode* q) {
        if (p == nullptr && q == nullptr) return true;
        if (p == nullptr || q == nullptr || p->val != q->val) return false;

        return (!isSameTree(p->left, q->left) || !isSameTree(p->right, q->right)) ? false : true;
    }
};